packages/block-library/src/navigation-link/edit.js

```diff
@@ -133,6 +133,82 @@ function getSuggestionsQuery( type, kind ) {
 	}
 }
 
+/**
+ * @typedef {'post-type'|'custom'|'taxonomy'} WPNavigationLinkKind
+ */
+
+/**
+ * Navigation Link Block Attributes
+ *
+ * @typedef {Object} WPNavigationLinkBlockAttributes
+ *
+ * @property {string}                [label]          Link text.
+ * @property {WPNavigationLinkKind}  [kind]           Kind is used to differentiate between term and post ids to check post draft status.
+ * @property {string}                [type]           The type such as post, page, tag, category and other custom types.
+ * @property {string}                [rel]            The relationship of the linked URL.
+ * @property {number}                [id]             A post or term id.
+ * @property {boolean}               [opensInNewTab]  Sets link target to _blank when true.
+ * @property {string}                [url]            Link href.
+ * @property {string}                [title]          Link title attribute.
+ */
+
+/**
+ * OnChange handler that updates block attributes when a link suggestion is chosen, a url is updated, or
+ * opensInNewTab is toggled
+ *
+ * @param {Object}                          updatedValue    New block attributes to update.
+ * @param {Function}                        setAttributes   Block attribute update function.
+ * @param {WPNavigationLinkBlockAttributes} blockAttributes Current block attributes.
+ *
+ */
+export const updateNavigationLinkBlockAttributes = (
+	updatedValue = {},
+	setAttributes,
+	blockAttributes = {}
+) => {
+	const {
+		label: originalLabel = '',
+		kind: originalKind = '',
+		type: originalType = '',
+	} = blockAttributes;
+	const {
+		title = '',
+		url = '',
+		opensInNewTab,
+		id,
+		kind: newKind = originalKind,
+		type: newType = originalType,
+	} = updatedValue;
+
+	const normalizedTitle = title.replace( /http(s?):\/\//gi, '' );
+	const normalizedURL = url.replace( /http(s?):\/\//gi, '' );
+	const escapeTitle =
+		title !== '' &&
+		normalizedTitle !== normalizedURL &&
+		originalLabel !== title;
+	const label = escapeTitle
+		? escape( title )
+		: originalLabel || escape( normalizedURL );
+
+	const type = newType === 'post_tag' ? 'tag' : newType.replace( '-', '_' );
```

Comment:
I feel we really need document what is happening here with `tag` and `post_tag`. It's happening all over the codebase now and it's a source of potential confusion.

Would an inline code comment here help?

I will look into writing a README around these things.

Reply:
Also let's document why we are swapping `-` with `_` with [references to trac](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L862)...etc.

Reply:
There's inconsistent usage of kebab case with the search rest API vs other server code, I'm not sure why hardcoded variations decided to pick "tag" over "post_tag", but possibly it was because the first pass may have tried to use it as a label display? https://github.com/WordPress/gutenberg/pull/24670/files#diff-7e333e3cb4e3e25a9d7349bcb6144c5af837ff2284bd98ceec2a616e40dd6213R66

One thing we could do is favor the actual entity name "post_tag", but also support the custom alias "tag" since persisted data is now mixed, depending on which WP version folks are using and if people left a tag link placeholder or not.

Reply:
I left a short comment https://github.com/WordPress/gutenberg/pull/31259/commits/c6a8a5984af5ec1f26af1fe61b9d7ebd698eacb3

I do think we can revisit the choice later.